rs/sns/governance/src/types.rs

```diff
@@ -2160,6 +2176,250 @@ impl NeuronParameters {
+    }
+
+    #[track_caller]
+    pub(crate) fn get_stake_e8s_or_panic(&self) -> u64 {
+        self.stake_e8s
+            .expect("Expected the stake_e8s to be present in NeuronRecipe")
+    }
+
+    #[track_caller]
+    pub(crate) fn get_neuron_id_or_panic(&self) -> &NeuronId {
+        self.neuron_id
+            .as_ref()
+            .expect("Expected NeuronId to be present in NeuronRecipe")
+    }
+
+    pub(crate) fn source_nns_neuron_id(&self) -> Option<u64> {
+        match &self.participant {
+            Some(Participant::NeuronsFund(nf)) => nf.nns_neuron_id.as_ref().cloned(),
+            _ => None,
+        }
+    }
+
+    #[track_caller]
+    pub(crate) fn construct_permissions_or_panic(
+        &self,
+        neuron_claimer_permissions: NeuronPermissionList,
+    ) -> Vec<NeuronPermission> {
+        self.construct_permissions(neuron_claimer_permissions)
+            .expect("Failed to construct permissions for neuron")
+    }
+
+    pub(crate) fn construct_permissions(
+        &self,
+        neuron_claimer_permissions: NeuronPermissionList,
+    ) -> Result<Vec<NeuronPermission>, String> {
+        let mut permissions = vec![];
+
+        let controller = self
+            .controller
+            .as_ref()
+            .ok_or("Expected controller to be present in NeuronRecipe".to_string())?;
+
+        permissions.push(NeuronPermission::new(
+            controller,
+            neuron_claimer_permissions.permissions,
+        ));
+
+        match &self.participant {
+            // NeuronsFund participants have additional permissions.
+            Some(Participant::NeuronsFund(neurons_fund_participant)) => {
+                let nns_neuron_controller = neurons_fund_participant.nns_neuron_controller.ok_or(
+                    "Expected the nns_neuron_controller to be present for NeuronsFundParticipant"
+                        .to_string(),
+                )?;
+                permissions.push(NeuronPermission::new(
+                    &nns_neuron_controller,
+                    vec![
+                        NeuronPermissionType::ManageVotingPermission as i32,
+                        NeuronPermissionType::SubmitProposal as i32,
+                        NeuronPermissionType::Vote as i32,
+                    ],
+                ));
+
+                for hotkey in neurons_fund_participant
+                    .nns_neuron_hotkeys
+                    .as_ref()
+                    .ok_or(
+                        "Expected the nns_neuron_hotkeys to be present for NeuronsFundParticipant"
+                            .to_string(),
+                    )?
+                    .principals
+                    .iter()
+                {
+                    permissions.push(NeuronPermission::new(
+                        hotkey,
+                        vec![
+                            NeuronPermissionType::SubmitProposal as i32,
+                            NeuronPermissionType::Vote as i32,
+                        ],
```

Comment:
I recommend lifting this set of permissions into a crate-level constant, e.g., at the top of this file.

Reply:
DOne